contract/models/contract.py

```diff
@@ -0,0 +1,516 @@
+                    field.readonly,
+                    field.company_dependent,
+                    field.name in self.NO_SYNC,
+                )
+            ):
+                if self.contract_template_id[field_name]:
+                    self[field_name] = self.contract_template_id[field_name]
+
+    @api.onchange("partner_id")
+    def _onchange_partner_id(self):
+        self.pricelist_id = self.partner_id.property_product_pricelist.id
+        self.fiscal_position_id = self.partner_id.property_account_position_id
+        if self.contract_type == "purchase":
+            self.payment_term_id = self.partner_id.property_supplier_payment_term_id
+        else:
+            self.payment_term_id = self.partner_id.property_payment_term_id
+        self.invoice_partner_id = self.partner_id.address_get(["invoice"])["invoice"]
+        return {
+            "domain": {
+                "invoice_partner_id": [
+                    "|",
+                    ("id", "parent_of", self.partner_id.id),
+                    ("id", "child_of", self.partner_id.id),
+                ]
+            }
+        }
+
+    def _convert_contract_lines(self, contract):
+        self.ensure_one()
+        new_lines = self.env["contract.line"]
+        contract_line_model = self.env["contract.line"]
+        for contract_line in contract.contract_line_ids:
+            vals = contract_line._convert_to_write(contract_line.read()[0])
+            # Remove template link field
+            vals.pop("contract_template_id", False)
+            vals["date_start"] = fields.Date.context_today(contract_line)
+            vals["recurring_next_date"] = fields.Date.context_today(contract_line)
+            new_lines += contract_line_model.new(vals)
+        new_lines._onchange_date_start()
+        new_lines._onchange_is_auto_renew()
+        return new_lines
+
+    def _prepare_invoice(self, date_invoice, journal=None):
+        """Prepare in a Form the values for the generated invoice record.
+
+        :return: A tuple with the vals dictionary and the Form with the
+          preloaded values for being used in lines.
+        """
+        self.ensure_one()
+        if not journal:
+            journal = (
+                self.journal_id
+                if self.journal_id.type == self.contract_type
+                else self.env["account.journal"].search(
+                    [
+                        ("type", "=", self.contract_type),
+                        ("company_id", "=", self.company_id.id),
+                    ],
+                    limit=1,
+                )
+            )
+        if not journal:
+            raise ValidationError(
+                _("Please define a %s journal for the company '%s'.")
+                % (self.contract_type, self.company_id.name or "")
+            )
+        invoice_type = "out_invoice"
+        if self.contract_type == "purchase":
+            invoice_type = "in_invoice"
+        move_form = Form(
```

Comment:
Is it really needed to use `odoo.tests.Form` in this situation? Shouldn't initializing a dictionary be more efficient?

Reply:
Yes, as this is not something about performance, but reliability: it tries to mimic how you input an invoice, for not having side effects of other modules that add some fields over invoices not being populated. Example: the payment mode added by `account_payment_partner` module. Not doing this forces us to make tricks like:

https://github.com/OCA/bank-payment/blob/12.0/account_payment_partner/models/account_invoice.py#L68-L84

Ideally, this wouldn't be needed, but with current ORM, this is the only way to ensure correctness. We can patch all possible extension modules, but that will be even worst. In a near future, when computed writable fields are totally implemented in core modules, we won't need this.